src/wp-includes/class-wp-theme.php

```diff
@@ -1576,6 +1576,11 @@ public function is_allowed( $check = 'both', $blog_id = null ) {
 	 * @return bool
 	 */
 	public function is_block_theme() {
+		if ( ! did_action( 'setup_theme' ) ) {
```

Comment:
Props to @peterwilsoncc for suggesting that we put a `_doing_it_wrong()` in this method rather than just in `wp_is_block_theme()`, since calling this method before themes are set up will have the same result.

Checking `did_action( 'setup_theme' )` rather than `did_action( 'after_setup_theme' )` is necessary, since this method is called from the constructor of this class whenever themes are being set up (via `wp_get_theme()` [called here](https://github.com/WordPress/wordpress-develop/blob/13fe8be4359222b7a53d4348dbd5fa81abe7080d/src/wp-settings.php#L689-L698)).

Reply:
> ... put a `_doing_it_wrong()` in this method rather than just in `wp_is_block_theme()`, since calling this method before themes are set up will have the same result.

My night brain was pondering whether we should put it in both. Doing so will be clearer for developers calling `wp_is_block_theme()` too early as it will show against the function they are calling.

If so, a shared string would be ideal with a placeholder for `__METHOD__` and `__FUNCTION__`.

Reply:
I'm not opposed, but I'm not sure triggering it in both places is necessary given that `wp_is_block_theme()` is only a thin wrapper for the `WP_Theme::is_block_theme()` method, which is what causes these kinds of errors. In either case, the developer would need to inspect the backtrace to see what function in their code triggered the error, and would see that they needed to make sure their code wasn't triggered until after the theme is set up.

Reply:
Yeah, let's not add it twice.